test(nns): Analyze list_neurons performance, and look for promising areas for optimization.

daniel-wong-dfinity-org opened the pull request:
# Usage

```
bazel test \
    --test_env=SSH_AUTH_SOCK \
    --test_output=streamed \
    --test_arg=--nocapture \
    //rs/nns/integration_tests:why_list_neurons_expensive \
&& cp -i \
  bazel-bin/rs/nns/integration_tests/test-127765624/why_list_neurons_expensive.runfiles/ic/list_neurons_*.svg \
  ./
```

(You need to be able to scp from dev@zh1-pyr07-zh.dfinity.network.)

# Limitations

1. The method that you want to measure must be an update, not a query. You can change a query to an update to measure it; however, make sure that you do **not** merge such changes, because they are very NOT backwards compatible!

2. You need to disable heartbeat. Otherwise, your flame graphs will have extraneous garbage in them.

# Next Steps

3. Prime Candid's cache by calling `ListNeuronsResponse::ty()` in `post_upgrade`, as suggested by Yan.

4. I'm not using all the space that I allocated to ic-wasm instrument. Before, when I tried to use all of it, I think it stomped on some other VirtualMemory's. (I guess the ideal solution is for ic-wasm instrument to support MemoryManager + VirtualMemory, rather than ask for raw stable memory space, but this might be difficult.)

5. Check the thread in the #dev channel where I am mainly talking to Yan Chen about getting this to work. I might be able to mine that thread for additional tasks to put into this "Next Steps" section...

6. Make as much of this sharable as possible, because every canister developer wants this (not just NNS).

# Analysis

One of the things that the `bazel test` command outputs is instructions vs. neurons. Using linear regression, I developed a very strong model (R^2 = 0.973):

```
Instructions = (1.03 * HeapNeurons + 5.08 StableMemoryNeurons + 47.5) * 1e5
```

This tells us that **the incremental cost of fetching and serving a neuron from heap is about 5x as much as it is from stable memory**.

This is also telling us that the **fixed cost/overhead of list_neurons is about 4.75 million instructions**.

See attached Jupyter notebook for source code: [list_neurons_instructions_vs_neuron_count.ipynb.txt][notebook]

[notebook]: https://github.com/user-attachments/files/17396897/list_neurons_instructions_vs_neuron_count.ipynb.txt

## Observations

When the model underestimates, the principal tends to have many neurons in stable memory. Likewise, when the model overestimates, the principal tends to have very few neurons in stable memory. Not sure why that is. This does not really make sense to me... (warning: I'm not a data scientist.)

### Anedotally

Seems like there can be BIG variance in the amount of instructions needed to load a neuron from stable memory. In some cases, it seems to take 100k instructions. In other cases, it seems to take 1.5M. It seems like what's happening here is that for "fast" stable memory neurons, theres no time spent fetching auxiliary data, but for "slow" stable memory neurons, most of the time fetching the neuron is spent loading auxiliary data. E.g. the neurons of principal ID an2fu-wmt2w-7fhvx-2g6qp-pq5pc-gdxu3-jfcr5-epiid-qhdzc-xnv3a-mae. We might want to dig into this, but I don't think there's time for that if we want to wrap up this investigation this week...

I think principal ID rxnbx-mhkez-7ckve-yboel-dnguy-dos76-hpdlg-fz5bf-xm5ou-72fre-wae is a good example of some (unsurprising) phenomena that seem to generalize:

1. A large chunk of instructions is spent in a couple of areas:
  1. Looking for the caller's neurons. This requires a range StableBTreeMap range scan. We maybe could speed this up if (some) of the index were in heap. It might make sense to split the index so that neurons in heap are **also indexed in heap**. However, this would introduce complexity.
  2. When a neuron is in stable memory, it takes much longer to load compared to heap (already noted earlier when discussing the linear model).
  3. Candid serialization of the response. This is significant when most of a principal's neurons are in heap. (This fact was hinted at earlier when the overhead of list_neurons was discussed in the context of the linear model.)

## Possible Solutions (not necessarily recommended!)

These three areas seem to be outside the control of the NNS. We maybe could speed up our use of stable memory by using a more efficient serialization scheme, but that would introduce more complexity.

When Rust + [WASM 64][64] support becomes mature, that might also help us avoid the cost of stable memory by keeping more data in heap. This has an important drawback though: we would have to "freeze dry" heap data when upgrading. If there is a large amount of this, upgrades become expensive. In any case, it does not seem like Rust + WASM 64 support is coming to the IC soon.

[64]: https://forum.dfinity.org/t/wasm64-beta-release-coming-soon/35791

# Examples

https://drive.google.com/file/d/1eBEuEKtKlmw4hpDkFYgrkX-h40jr9s_H/view?usp=sharing

💡 You will have to download this in order to be able click on boxes in this example ☝️ to zoom in on those boxes, because the Google Drive viewer does not seem to support this.

## Related Work

[Max's range_neurons optimization][zip]. IIUC, this is currently only used in background (data validation) tasks, so it would not help us here.

[zip]: https://github.com/dfinity/ic/pull/2033

daniel-wong-dfinity-org committed:
Trying to generate flame graph(s) for list_neurons (in NNS governance). Running into a problem where ic-wasm instrument uses the 32-bit stable memory API, but we are already using more than 4 GiB of stable memory. Yan Chen confirmed that ic-wasm does not yet support the 32-bit API, but that can be fixed pretty easily.

daniel-wong-dfinity-org committed:
Upgrade our version of ic-wasm to 0.9.0, because the previous version does not support the 64-bit stable memory API (the 32-bit API is deprecated).

This required us to upgrade our version of walrus from 0.21.1 to 0.22.0. Not sure yet if this has bad side effects...

daniel-wong-dfinity-org committed:
Mostly, misc comment changes. The only "real" code change is switching from more lax from_utf8_lossy to stricter from_utf8 + unwrap.

basvandijk commented on bazel/external_crates.bzl:
```diff
@@ -648,7 +651,9 @@ def external_crates_repository(name, cargo_lockfile, lockfile, sanitizers_enable
             ),
             "ic-wasm": crate.spec(
-                version = "^0.8.4",
+                git = "https://github.com/dfinity/ic-wasm", # DO NOT MERGE
+                rev = "e6b26402f4aa053903ffe4cb3ba2dac5f9dca89b", # DO NOT MERGE
+                # DO NOT MERGE version = "^0.8.4",
```
Is this PR meant to be a draft or should IDX review this? If the latter, why the `DO NOT MERGE` comments?

basvandijk replied:
Ok I see you converted the PR to draft.

daniel-wong-dfinity-org replied:
This is draft.

I took it out of draft briefly in the hopes that it would cause the tests to actually run.

I tried to remove reviewers, but it won't let me.

daniel-wong-dfinity-org committed:
Hack to work around reading too far into stable memory, apparently be ic-wasm instrument 😡

daniel-wong-dfinity-org committed:
lint

daniel-wong-dfinity-org committed:
lint

daniel-wong-dfinity-org committed:
Getting flame graphs for the first time. This required a few orthogonal changes:

1. start_address is a byte offset, not page.
2. Read "name" metadata from ic-wasm instrumented WASM. This is later used to visualize.
3. Visualize. Output is at list_neurons.svg. This involved copying fn render_profiling from the ic-repl repo.

(I will upload an example list_neurons.svg to GitHub shortly.)

daniel-wong-dfinity-org commented on rs/nns/integration_tests/src/why_list_neurons_expensive.rs:
```diff
@@ -0,0 +1,253 @@
+    )
+    .unwrap();
+    println!("");
+    println!(
+        "custom section 'name' payload:\n{:#?}",
+        name_custom_section_payload
+    );
```
This only prints 4 items:

```
custom section 'name' payload:
{
    3921: "__get_profiling",
    3922: "__get_cycles",
    3923: "__toggle_tracing",
    3924: "__toggle_entry",
}
```

daniel-wong-dfinity-org replied:
This is solved by keep_name_section = True in BUILD file.

daniel-wong-dfinity-org committed:
In order for names to be in flame graphs, we need to add `keep_name_section = True` to our rust_canister rules in BUILD files.

daniel-wong-dfinity-org committed:
Print list_neurons result.

daniel-wong-dfinity-org committed:
Find top principals by neuron count. This required some visibility hacking.

daniel-wong-dfinity-org committed:
squelch some lints

daniel-wong-dfinity-org committed:
Comment out some prints.

daniel-wong-dfinity-org committed:
Generate many flame graphs to see how list_neuron scales.

daniel-wong-dfinity-org commented on rs/nns/governance/canister/canister.rs:
```diff
@@ -497,6 +508,38 @@ async fn claim_or_refresh_neuron_from_account(
+/// controller of N, a hotkey (or both).
+use std::collections::HashMap;
+#[update]
+fn principal_to_neuron_count() // DO NOT MERGE
```
principal_to_neuron_count

daniel-wong-dfinity-org committed:
Measure more list_neuron calls. In particular, try principals that have fewer (more typical) number of neurons.

daniel-wong-dfinity-org committed:
Overhaul comments.

daniel-wong-dfinity-org committed:
Gather more performance data about list_neurons.

daniel-wong-dfinity-org committed:
Break out (per principal) neuron count into heap neuron count vs. stable memory neuron count (again, per principal).

daniel-wong-dfinity-org committed:
Misc comment changes.

daniel-wong-dfinity-org committed:
Bug: principal's neuron counters were not being incremented.

daniel-wong-dfinity-org committed:
CSV instead of TSV.

daniel-wong-dfinity-org committed:
Misc cosemetic changes. E.g. comments, reflow, debug prints, wording. Some unused imports, but I think everything builds at this point.

My favorite change here: do not tell inferno how wide to make the svg. This is super nice, because then it is displayed using as much horizontal space as you give it.

daniel-wong-dfinity-org committed:
Try to prime the Candid cache. Yan recommended this, but it doesn't seem to help. Not sure why.

daniel-wong-dfinity-org committed:
Spin out function that converts from ic wasm's trace format into the format used by inferno. Includes tests that show behavior is preserved. (The next commit is going to get rid of the original impl, which was copy n' pasted, and not very self-explanatory.)

daniel-wong-dfinity-org committed:
Deleted old implementation of conversion from the trace/profile format used by ic-wasm to the format used by inferno. Previous commit shows that the new impl is equivalent.

Commit:
Automatically updated Cargo*.lock

daniel-wong-dfinity-org committed:
small fixes

daniel-wong-dfinity-org committed:
Moved some code.

daniel-wong-dfinity-org committed:
Clarify error string.

daniel-wong-dfinity-org committed:
Moved more code.

daniel-wong-dfinity-org committed:
Allocate profiling space on install, not just upgrade.

daniel-wong-dfinity-org commented on rs/nns/integration_tests/src/why_list_neurons_expensive.rs:
```diff
@@ -0,0 +1,417 @@
+#[test]
+fn test_why_list_neurons_expensive() {
+    // Load golden nns state into a StateMachine.
+    let state_machine = new_state_machine_with_golden_nns_state_or_panic();
```
This is not necessary for flamegraphing, but many ppl trying to generate flamegraphs will find this beneficial, because this loads realistic data.

daniel-wong-dfinity-org commented on rs/nns/integration_tests/src/why_list_neurons_expensive.rs:
```diff
@@ -0,0 +1,417 @@
+    // that can be used to generate a flame graph. This space will be used by
+    // ic-wasm instrument.
+
+    println!("\nAllocating stable memory for profiling...\n");
```
This is where we begin "instrumenting" the WASM, i.e. inject code that measures running cost. Measurements are saved during execution, and fetched afterwards. Then, they are fed into flamegraph generation.

daniel-wong-dfinity-org commented on rs/nns/integration_tests/src/why_list_neurons_expensive.rs:
```diff
@@ -0,0 +1,417 @@
+    println!("");
+    println!("Done installing governance WITH ic-wasm profiling ENABLED.");
+    println!("Ready for fine-grained performance measurement 👍");
+    println!("");
```
End instrumentation.

daniel-wong-dfinity-org commented on rs/nns/integration_tests/src/why_list_neurons_expensive.rs:
```diff
@@ -0,0 +1,417 @@
+    println!("Ready for fine-grained performance measurement 👍");
+    println!("");
+
+    // Grab a sample of principals that have "associated" neurons. More
+    // precisely, these principals either control or are a hotkey of some
+    // (nonzero number of) neurons. These will later be used as callers of the
+    // list_neurons method.
+    let principal_id_to_neuron_count = state_machine
+        .execute_ingress(
+            GOVERNANCE_CANISTER_ID,
+            "principal_id_to_neuron_count",
+            Encode!().unwrap(),
+        )
+        .unwrap();
+    let principal_id_to_neuron_count = match principal_id_to_neuron_count {
+        ic_types::ingress::WasmResult::Reply(ok) => ok,
+        doh => panic!("{:?}", doh),
+    };
+    let principal_id_to_neuron_count = Decode!(
+        &principal_id_to_neuron_count,
+        Vec<(PrincipalId, (u64, u64))>
+    )
+    .unwrap();
+    println!("");
+    println!("principal_id_to_neuron_count:");
+    println!(
+        "{}",
+        clamp_debug_len(&principal_id_to_neuron_count, 1 << 10)
+    );
+    println!("");
```
This is not of general interest to ppl who want to make their own flamegraphs. Rather, this is just to figure out what data to feed to the code that we are about to measure.

daniel-wong-dfinity-org commented on rs/nns/integration_tests/src/why_list_neurons_expensive.rs:
```diff
@@ -0,0 +1,417 @@
+    println!("");
+
+    // Measure a variety of list_neuron calls.
+    for (caller, (heap_neuron_count, stable_memory_neuron_count)) in principal_id_to_neuron_count {
+        // Skip super heavy principals, because list_neurons does not even work for them.
+        let is_principal_too_heavy = [
+            PrincipalId::from_str("dies2-up6x4-i42et-avcop-xvl3v-it2mm-hqeuj-j4hyu-vz7wl-ewndz-dae").unwrap(),
+            PrincipalId::from_str("renrk-eyaaa-aaaaa-aaada-cai").unwrap(),
+        ].contains(&caller);
+        if is_principal_too_heavy {
+            continue;
+        }
+
+        // Call code being measured.
+        #[allow(unused)] // DO NOT MERGE
+        let result = list_neurons(
+            &state_machine,
+            caller,
+            ListNeurons {
+                include_neurons_readable_by_caller: true,
+                include_public_neurons_in_full_neurons: Some(false),
+                include_empty_neurons_readable_by_caller: Some(true),
+                neuron_ids: vec![],
+                neuron_subaccounts: None,
+                page_number: None,
+                page_size: None,
+            },
+        );
```
Call code that we want to measure/optimize.

daniel-wong-dfinity-org commented on rs/nns/integration_tests/src/why_list_neurons_expensive.rs:
```diff
@@ -0,0 +1,417 @@
+        */
+
+        // Fetch measurement.
+        let profiling = primitive_get_profiling(&state_machine, GOVERNANCE_CANISTER_ID);
```
Fetch measurement(s).

daniel-wong-dfinity-org commented on rs/nns/integration_tests/src/why_list_neurons_expensive.rs:
```diff
@@ -0,0 +1,417 @@
+        // Fetch measurement.
+        let profiling = primitive_get_profiling(&state_machine, GOVERNANCE_CANISTER_ID);
+
+        // Visualize where instructions get consumed.
+        let title = format!(
+            "{}/{} neurons\n\
+             {}\n\
+             Candid cache primed
+             no width",
+            heap_neuron_count, stable_memory_neuron_count, caller,
+        );
+        let destination = format!(
+            "list_neurons_no_width_{:0>4}_heap_neurons_{:0>4}_stable_memory_neurons_{}.svg",
+            heap_neuron_count, stable_memory_neuron_count, caller,
+        );
+        // I guess this is so that bars are labeled with names, not opaque IDs.
+        // See https://sourcegraph.com/github.com/dfinity/ic-repl@746bea25ddd4cc98709f6b9eaa283f32a21ac30d/-/blob/src/helper.rs?L504
+        let name_custom_section_payload = Decode!(
+            &read_custom_section(&instrumented_governance_wasm, "icp:public name"),
+            BTreeMap<u16, String>
+        )
+        .unwrap();
+        let cost = render_profiling( // DO NOT MERGE
+            profiling,
+            &name_custom_section_payload,
+            &title,
+            PathBuf::from(destination),
+        )
+        .unwrap();
```
Generate flamegraph/visualization from measurement(s).

daniel-wong-dfinity-org commented on rs/nns/test_utils/src/state_test_helpers.rs:
Low level measurement parsing helpers.

daniel-wong-dfinity-org reviewed (commented):
diff reading guide. As you maybe inferred from the PR description, the main file is `why_list_neurons_expensive.rs`.
